### PR TITLE
Issue #15541: Add new isOfType that accepts List of Integers as parameter

### LINKED ISSUE #15541: New isOfType Method that Accepts List as its Parameter
The `isOfType` method currently accepts an array of `int` as its second parameter. We should overload this method to accept a `List<Integer>`.

https://github.com/checkstyle/checkstyle/blob/8f856ecb1f2182d0b20fd22a83be115c1577e49f/src/main/java/com/puppycrawl/tools/checkstyle/utils/TokenUtil.java#L285-L293

Reason: 
detected at https://github.com/checkstyle/checkstyle/pull/15505#discussion_r1725332873.
If we have `int[]` as a class field the `ImmutabilityTest` will fail as final arrays are technically mutable so another choice is to use Lists to satisfy the `ImmutabilityTest`.

### PR DESCRIPTION
**Pull request by ahmedyoussefg**
Closes #15541

### CODE REVIEW

**Comment by romani** [comment]
Thanks a lot for helping us.
Please use new method at https://github.com/checkstyle/checkstyle/blob/9e7ad9e42740346ff9a8a1dfecae8d37d40df43a/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MissingSwitchDefaultCheck.java#L179

This array should be list and should be immutable list that is created ones and stored as class field.

**Comment by romani** [comment]
Before pushing code , please run `mvn clean verify ` on your local , and fix all errors if any appears.

**Comment by mahfouz72** [comment]
@romani  Do you have an idea why we lost some checks here? we should have ~112

File: src/main/java/com/puppycrawl/tools/checkstyle/utils/TokenUtil.java
Comment (mahfouz72) [inline src/main/java/com/puppycrawl/tools/checkstyle/utils/TokenUtil.java:338]: > ast != null

This condition will not get covered by real check execution so pitest will complain. The real usage of this method in the checks always sends a `ast` that is always not null so this mutation will survive.

@romani  Do you agree with having a pure JUnit for this case like in: 

https://github.com/checkstyle/checkstyle/blob/15de92ae48ddea70c478d948bf2258906eb34b74/src/test/java/com/puppycrawl/tools/checkstyle/utils/TokenUtilTest.java#L362-L383


This condition is just a safety from NPE but there is no real execution case where we send a null to this method. What do you think?

**Comment by romani** [comment]
New contributor PRs are not exciting all , until I approve run. After few approvals, GitHub trust new contributors

File: src/main/java/com/puppycrawl/tools/checkstyle/utils/TokenUtil.java
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/utils/TokenUtil.java:338]: No pure unit testing.
All through Input files.

Comment (ahmedyoussefg) [reply src/main/java/com/puppycrawl/tools/checkstyle/utils/TokenUtil.java:338]: I want to confirm my understanding of the issue. The problem with the changes I made is that the `ast != null` condition is not covered during either JUnit tests or normal execution. I considered adding something like the following in the `public void testIsOfTypeFalse()` method:

```java
final DetailAstImpl astForTest3 = null; 
List<Integer> typesList = List.of(TokenTypes.LITERAL_FOR, TokenTypes.LITERAL_IF, TokenTypes.LITERAL_ELSE);
final boolean result4 = TokenUtil.isOfType(astForTest3, typesList);
assertWithMessage("Token type should not match") 
    .that(result4) 
    .isFalse(); 
```
However, since you mentioned avoiding pure unit testing, should I think of a different approach to ensure the `ast != null` condition is covered?

Comment (mahfouz72) [reply src/main/java/com/puppycrawl/tools/checkstyle/utils/TokenUtil.java:338]: Then there will not be a nullity check in this method. However,  This will always be checked by the caller (We are sure that the `ast` passed in this method will not be null). So IMO,  yes we can only have `public static boolean isOfType(int type, List<Integer> types)`.

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/utils/TokenUtil.java:338]: We are not a Library for other on universal methods. Better to keep code that we know why it exists. It in future some Check would benefit from null checking, we will add such method and it will naturally cover such edge cases.

Comment (ahmedyoussefg) [reply src/main/java/com/puppycrawl/tools/checkstyle/utils/TokenUtil.java:338]: We can modify the return value of `isSwitchExpression` to: 
```java
return !TokenUtil.isOfType(ast.getParent().getType(), switchStatementParents);
```
If we are certain that the parent of `ast` is always not null, otherwise, I don't currently see a way to avoid the null check

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/utils/TokenUtil.java:338]: If ast is not null, let's not make check for it.
Code coverage will demand coverage for all.

File: src/main/java/com/puppycrawl/tools/checkstyle/utils/TokenUtil.java
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/utils/TokenUtil.java:337]: Please remove this method .
No need in this abstraction layer.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MissingSwitchDefaultCheck.java
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MissingSwitchDefaultCheck.java:181]: Move it class field

**Review by romani — CHANGES_REQUESTED** [review]
Items

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MissingSwitchDefaultCheck.java
Comment (ahmedyoussefg) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MissingSwitchDefaultCheck.java]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/utils/TokenUtil.java
Comment (ahmedyoussefg) [reply src/main/java/com/puppycrawl/tools/checkstyle/utils/TokenUtil.java]: Done

**Review by romani — APPROVED** [review]
Ok to merge if CI pass.
New regression report is required.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MissingSwitchDefaultCheck.java
Comment (mahfouz72) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MissingSwitchDefaultCheck.java:95]: `Parents of switch statement`. => `Represents the possible parent tokens of a switch statement`

**Review by mahfouz72 — CHANGES_REQUESTED** [review]
Minor to improve Javadoc clarity.

**Comment by mahfouz72** [comment]
@ahmedyoussefg  Please see [workflows readme](https://github.com/checkstyle/checkstyle/blob/master/.github/workflows/README.md#regression-report-workflow---regression-reportyml) to learn how to generate regression report

**Comment by ahmedyoussefg** [comment]
Github, generate report for MissingSwitchDefault/all-examples-in-one

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MissingSwitchDefaultCheck.java
Comment (ahmedyoussefg) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MissingSwitchDefaultCheck.java]: Done

**Review by mahfouz72 — APPROVED** [review]
Thanks a lot!!

**Comment by mahfouz72** [comment]
@ahmedyoussefg we need to add this check tests as a target tests in pitest utils profile in pom.xml

@romani Ideally we should delete UtilTest classs as all of them are pure junit and add all check tests as a target tests in pitest-utils profile

**Comment by romani** [comment]
We are in transition from pure unit testing to BDD. Pure unit testing is allowed but requires special comment to explain why BDD is not possible.
But a lot needs to be refactored to remove examples of bad testing approaches.

File: pom.xml
Comment (ahmedyoussefg) [inline pom.xml:4996]: like this ? @mahfouz72

Comment (mahfouz72) [reply pom.xml:4996]: yes pitest should be green now.
@romani please approve other workflows

Comment (romani) [reply pom.xml:4996]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MissingSwitchDefaultCheck.java
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MissingSwitchDefaultCheck.java:96]: Lets convert this type to Set. We just do search over this collection, so it will be slightly quicker.

**Review by romani — CHANGES_REQUESTED** [review]
Item

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MissingSwitchDefaultCheck.java
Comment (ahmedyoussefg) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MissingSwitchDefaultCheck.java]: Done

**Review by romani — APPROVED** [review]

**Comment by romani** [comment]
@ahmedyoussefg, thanks a lot for your contribution.
Please do more.
I highly recommend to start introduction to our project by fixing issues with labels "good xxxx issue".

**Pull request merged**
